### PR TITLE
DO NOT MERGE chore: Refactor

### PR DESCRIPTION
**Pull request by iAmMichaelConnor**
# Description

I spent a bit of time refactoring BigNum into what I hope is an easier-to-follow design. 

I haven't touched any of the mathematics; just moved functions around.

Things I sought to improve:

- The BigNum and RuntimeBigNum were somewhat cyclic structs: their methods called each-others' methods. The comments in the code even mentioned this being problematic and confusing. It also meant if you tried to introduce a new method (like I tried), you could accidentally fall afoul of the cyclic-ness.
    - To solve this, I've put made all of the proper "impl" functions as "free functions" instead of methods. Both structs still exists. It's just their methods are now clean wrappers which call the free functions. This way, both structs can call those methods without any cyclic dependencies on each-other. In fact, the structs now don't know about each other at all.
- When creating a BigNum from a BigNumInstance, many of the methods of BigNum weren't available (I think because of the above cyclic problems), and the pretty operators (+,-,*,/) weren't available. 
    - I think this is because really, runtime bignum is a distinct thing that deserves its own struct. So, I introduced "RuntimeBigNum", which has exactly the same methods and pretty operators as BigNum. The only difference is whether the params are known at compile-time or not.
- The above simplifications meant I could remove lots of traits and generic parameters which made the code difficult to follow. Hopefully it reads more cleanly.
- Broke big files into a few smaller files.
- Removed the "dummy struct" that needed to be declared when defining a new set of runtime bignum params.

Anyway, all the tests pass, wooh!

Lmk what you think. :)

### CODE REVIEW

**Comment by iAmMichaelConnor** [comment]
Looks like nightly is happy, but 0.34.0 isn't. Will investigate.

**Comment by iAmMichaelConnor** [comment]
Before: 
image

After:
image

**Comment by iAmMichaelConnor** [comment]
If approved, I'll do the README

File: src/utils/map.nr [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Copied from std::array, because I couldn't figure out how to import the `map` method of the weird trait for an array.
+// And the reason I wanted direct access to it, is because I couldn't figure out how to implement a double map.
+pub fn map<T, let N: u32, U, Env>(arr: [T; N], f: fn[Env](T) -> U) -> [U; N] {
```
Comment (TomAFrench) [inline src/utils/map.nr:3]:
```suggestion
pub(crate) fn map<T, let N: u32, U, Env>(arr: [T; N], f: fn[Env](T) -> U) -> [U; N] {
```

File: src/runtime_bignum.nr [hunk 2]
```diff
@@ -1,1551 +1,344 @@
...
+use crate::expressions::{__compute_quadratic_expression, evaluate_quadratic_expression};
+use crate::serialization::{from_be_bytes, to_le_bytes};
+
+pub struct RuntimeBigNum<let N: u32> {
+    limbs: [Field; N],
+    params: BigNumParams<N>,
 }
 
-impl<let N: u32, Params> BigNumInstanceTrait<BigNum<N, Params>> for BigNumInstance<N, Params> where Params: BigNumParamsTrait<N> {
+impl<let N: u32> RuntimeBigNum<N> {}
```
Comment (TomAFrench) [inline src/runtime_bignum.nr:21]:
```suggestion
```

**Review by TomAFrench — APPROVED** [review]

**Comment by iAmMichaelConnor** [comment]
See here for a `main()` function which calls bignum, as evidence that the constraint counts are unchanged: https://github.com/iAmMichaelConnor/debug-bignum-refactor

File: src/bignum.nr [hunk 3]
```diff
@@ -216,6 +218,19 @@ impl<let N: u32, Params> BigNumTrait<N> for BigNum<N, Params> where Params: BigN
         __batch_invert_slice(params, x.map(|bn| Self::get_limbs(bn))).map(|limbs| Self { limbs })
     }
 
+    unconstrained fn __tonelli_shanks_sqrt(self) -> std::option::Option<Self> {
+        let params = Params::get_params();
+        let maybe_limbs = unsafe {
+            __tonelli_shanks_sqrt(params, self.limbs)
+        };
+        let result: std::option::Option<Self> = if maybe_limbs.is_some() {
```
Comment (TomAFrench) [inline src/bignum.nr:226]: This could be done with `.map`

**Comment by iAmMichaelConnor** [comment]
@TomAFrench check out the strangeness of my latest commit diff: https://github.com/noir-lang/noir-bignum/pull/27/commits/33611026357627f3e11005120e14b3f271c1b8e2

The test of `assert(x == x)` was failing, because `eq` makes a call to `assert_max_bit_size()` on a value derived from `params.modulus_bits_getter()`. I was hoping this getter function would always return a value that's recognised as constant, because it is always a function of the form `|| 255` (or some other constant number).

Anyway, it seems that calling `x.__tonelli_shanks_sqrt()` was confusing the compiler into thinking the modulus_bits wasn't constant. I suspected it's because the function was returning an Option. Changing from an option to an ugly tuple fixed the compiler error. Not sure why though.

**Comment by TomAFrench** [comment]
Sorry I'm on a plane :shrug:

**Pull request closed without merging**